library/alloc/src/macros.rs

```diff
@@ -99,9 +108,16 @@ macro_rules! vec {
 /// format!("hello {}", "world!");
 /// format!("x = {}, y = {y}", 10, y = 30);
 /// ```
+#[allow_internal_unstable(display_type_guard)]
 #[macro_export]
 #[stable(feature = "rust1", since = "1.0.0")]
 macro_rules! format {
+    // A faster path for simple format! calls.
+    ("{}", $($arg:tt,?)+) => {{
+        let fast = |t: NeedsDisplay<_> | $crate::string::ToString::to_string(t.inner);
+        // This TokenTree must resolve to a Displayable value to be valid.
+        fast(NeedsDisplay { inner: &{$($arg)*} })
+    }};
```

Comment:
We will want a codegen test I think for this, at least -- it looks like this is just not triggering to me, because I would expect `NeedsDisplay` to not resolve here. Generally we need the type to be public in order for it to be constructible outside the current crate.

Reply:
> codegen – tests that compile and then test the generated LLVM code to make sure that the optimizations we want are taking effect. See LLVM docs for how to write such tests.
Ah, one of these? Well, that just made this PR more Exciting.

Reply:
Note that before working on that I would really fix this macro to make sure this arm is getting expanded, because right now that seems pretty unlikely to me. See also the comment below about needing :ident on the arm, most likely, which'll further simplify this branch.